### ISSUE TITLE
pephub Device Authorization Grant

### ISSUE THREAD

**Issue by khoroshevskyi** [pepkit/pephub#115]
I was thinking about pephubclient authorization, and In my opinion the best solution will be creating ~the authorization code grant~ Device Authorization Grant. We can implement this strategy, similar way how github does, or microsoft...
https://www.rfc-editor.org/rfc/rfc8628

For fastapi there is discussion on this topic: https://github.com/tiangolo/fastapi/discussions/9141

This isn't the easiest solution, but in my opinion, is the best for future pephub development

**Comment by nleroy917** [comment]
I'm pretty sure I've already implemented a basic auth code flow. This is the current architecture:

<img width="1507" alt="image" src="https://user-images.githubusercontent.com/41063083/224840088-5879ae37-25d1-48aa-80bb-d6a531a9ad40.png">

**Comment by nleroy917** [comment]
After digging into the depths of the Microsoft azure cli, I found [this file](https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/bd931a22e80381cb410d45c8341c40d49a80c3d6/msal/oauth2cli/authcode.py#L4) which I believe is doing precisely what @nsheff suggested today: Start a miniature web server, redirect there and wait for an auth code response. You might even be able to do this with standard python libraries:

```python
import http.server
import socketserver
import urllib.parse

class MyHandler(http.server.SimpleHTTPRequestHandler):
    def do_GET(self):
        parsed_url = urllib.parse.urlparse(self.path)
        if parsed_url.path == "/callback":
            code = urllib.parse.parse_qs(parsed_url.query)["code"][0]
            self.send_response(200)
            self.send_header("Content-type", "text/html")
            self.end_headers()
            self.wfile.write(b"<html><body><h1>Authentication successful!</h1></body></html>")
            self.server.auth_code = code
        else:
            super().do_GET()

with socketserver.TCPServer(("", 8000), MyHandler) as httpd:
    print("Listening on port 8000...")
    httpd.serve_forever()
```

**Comment by khoroshevskyi** [comment]
oops, I sent incorrect link: https://www.rfc-editor.org/rfc/rfc8628, so this is  "for Internet-
   connected devices that either lack a browser to perform a user-agent-
   based authorization or are input constrained to the extent that
   requiring the user to input text in order to authenticate during the
   authorization flow is impractical. " 

The same way how github is doing it

**Comment by khoroshevskyi** [comment]
Additionally I found interesting explanation of oauth2: https://dev.to/ramushetty/oauth2-authorization-grants-313i
![image](https://res.cloudinary.com/practicaldev/image/fetch/s--GPuNEMNs--/c_limit%2Cf_auto%2Cfl_progressive%2Cq_auto%2Cw_880/https://dev-to-uploads.s3.amazonaws.com/i/o8ajaelpjcmxtfnlvohp.png)

**Comment by nleroy917** [comment]
### Notes from March 14th, 2023
- We will add device authentication flow to pephub

**Base endpoint:** `/auth/device`

We have four separate endpoints:
1. `POST /init`: instantiate the login flow, mints a code, returns with code and/or URL (this token should be only good for certain amount of time)
2.  `GET /login/<device-code>`: via the browser, open this URL and grab token that was just minted from the path, push user through GitHub login. Mint a **new** token that is associated with this device code, send it through the state to github (`json.dumps(state)`).
3. `GET /login/callback`: server endpoint that GitHub redirects to.. should read the random temporary device-code identifier and exchange it for the device code stored in memory... then mint a JWT and associate with the actual device code.
4. `POST /token`: finally, exchange the original device code for the actual JWT via a `POST` request

**Comment by nleroy917** [comment]
This is solved in the latest release

**Comment by nleroy917** [comment]
This is now available in the latest release (`0.7.x`)
